rfd/0182-multi-port-tcp-app-access.md

```diff
@@ -0,0 +1,544 @@
+and the cert is the same for all connections coming from a single local proxy.
+
+The only thing that changes for multi-port apps is that we add `app_target_port` field which
+includes the port to which the connection was forwarded to. The session ID is the same for
+connections forwarded to the same port using the same cert and different for connections on
+different ports. This means that two connections on two different ports for the same app are going
+to generate two `app.session.start` events which are identical, with the exception of
+`app_target_port`, `time`, `sid`, and `uid` fields.
+
+### Observability
+
+Whether the target port makes it all the way from the client to the app service can be observed by
+the `app.session.start` audit event. It should include the correct port in the `app_target_port`
+field.
+
+We don't expect the implementation to impact performance. The cert used to transport the port number
+already includes routing details for apps. The app service already has access to the app spec to
+verify whether the target port is valid, so it doesn't need to perform any additional network calls.
+
+### Product Usage
+
+The telemetry already tracks TCP app sessions through the `tp.session.start.v2` event where
+`tp.session_type` is set to `app_tcp`. We're going to extend this event with a new boolean property
+called `tp.app.is_multi_port`, which value is going to depend on the presence of `AppTargetPort` in
+`AppMetadata` of `apievents.AppSessionStart`.
+
+### Proto Specification
+
+Changes in `api/proto/teleport/legacy/types/types.proto`:
+
+```diff
+ message AppSpecV3 {
+ …
+   // Only applicable to AWS App Access.
+   // If present, the Application must use the Integration's credentials instead of ambient credentials to access Cloud APIs.
+   string Integration = 9 [(gogoproto.jsontag) = "integration,omitempty"];
++  // Ports is a list of ports and port ranges that an app agent can forward connections to.
++  // Only applicable to TCP App Access.
++  // If this field is not empty, URI is expected to contain no port number and start with the tcp
++  // protocol.
++  // If a value within this field is a port range, the start of the range must be less than the end
++  // of the range. Any individual port cannot fall outside of the 1-65535 range.
++  repeated PortRange Ports = 10 [(gogoproto.jsontag) = "ports,omitempty"];
++}
++
++// PortRange describes a port range used for in TCP apps. It can be used to describe a single port
++// in which case the Start field is the port and the End field is 0.
++message PortRange {
++  // Start describes the start of the range. It must be between 1-65535.
++  uint32 Start = 1 [(gogoproto.jsontag) = "start"];
++  // End describes the end of the range, inclusive. It must be between 2-65535 and be greater than
++  // Start when describing a port range. When describing a single port, it must be set to 0.
```

Comment:
I would remove the special case of 0 and require the `[start, end]` inclusive interval to be nonempty and contained in `[1, 65535]` (which means `1 <= start <= end <= 65535`, with single ports resulting in `start == end` but requiring no extra code).

Then again, kubernetes networkpolicies use something like the following, so there _is_ prior art to having a zero/empty/missing end value to define a single port, so I am not super attached to the suggestion of always requiring a range.
```yaml
ports:
  - protocol: TCP
    port: 1234
  - protocol: TCP
    port: 32000
    endPort: 32768
```

Reply:
Do we want to disallow or implicitly convert apps that have one single-port range into a standard single-port app? Do we want to push users to define apps as multi-port even when they're only using a single port?

Reply:
i kinda like port/endPort because "start" is pretty awkward on its own
```yaml
ports:
  - start: 1234
  - start: 2345
```
feels worse than
```yaml
ports:
  - port: 1234
  - port: 2345
```

Reply:
Yeah, as discussed on Slack, I'll have to redo how the ports are defined because it seems like my initial idea here (`ports: [4080-4090, 4095]`) requires too much effort to be feasible at the moment.

> i kinda like port/endPort because "start" is pretty awkward on its own

Agree, that's why I'm partial to following what k8s does too. I know Edoardo mentioned on Slack how people who use automation to set up apps are not going to care about shorthands, but I feel like port/endPort strikes a good balance between first-time experience of trying out stuff by hand, setting up automation around it and writing code for (un)marshalling and validation.

> Do we want to disallow or implicitly convert apps that have one single-port range into a standard single-port app? Do we want to push users to define apps as multi-port even when they're only using a single port?

That's an interesting question. I'll think about it and come back to it tomorrow.